File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/error/MaplibreStatus.java
package org.maplibre.nativeffi.error;

/** Status categories reported by the native Maplibre C ABI. */
public enum MaplibreStatus {
  OK(0),
  INVALID_ARGUMENT(-1),
  INVALID_STATE(-2),
  WRONG_THREAD(-3),
  UNSUPPORTED(-4),
  NATIVE_ERROR(-5),
  UNKNOWN(Integer.MIN_VALUE);

  private final int nativeCode;

  MaplibreStatus(int nativeCode) {
    this.nativeCode = nativeCode;
  }

  /** Returns the C ABI status value for known statuses. */
  public int nativeCode() {
    return nativeCode;
  }

  public static MaplibreStatus fromNative(int nativeCode) {
    return switch (nativeCode) {
      case 0 -> OK;
      case -1 -> INVALID_ARGUMENT;
      case -2 -> INVALID_STATE;
      case -3 -> WRONG_THREAD;
      case -4 -> UNSUPPORTED;
      case -5 -> NATIVE_ERROR;
      default -> UNKNOWN;
    };
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/error/MaplibreStatus.kt
package org.maplibre.nativeffi.error

import kotlin.jvm.JvmInline

/** Status categories reported by the native MapLibre C ABI. */
@JvmInline
public value class MaplibreStatus(public val nativeCode: Int) {
  public companion object {
    public val OK: MaplibreStatus = MaplibreStatus(0)
    public val INVALID_ARGUMENT: MaplibreStatus = MaplibreStatus(-1)
    public val INVALID_STATE: MaplibreStatus = MaplibreStatus(-2)
    public val WRONG_THREAD: MaplibreStatus = MaplibreStatus(-3)
    public val UNSUPPORTED: MaplibreStatus = MaplibreStatus(-4)
    public val NATIVE_ERROR: MaplibreStatus = MaplibreStatus(-5)

    /** Returns the Kotlin status category for a C ABI status value. */
    internal fun fromNative(nativeCode: Int): MaplibreStatus = MaplibreStatus(nativeCode)
  }
}
